File: dart/day10_part2_2020.dart
import 'dart:io';

void main() {
  var file = File('input.txt');
  var lines = file.readAsLinesSync();

  var adapters = [0];
  for (var line in lines) {
    adapters.add(int.parse(line));
  }

  adapters.sort();
  adapters.add(adapters.last + 3);

  print(countArrangements(adapters));
}

int countArrangements(List<int> adapters) {
  var ways = {0: 1};

  for (var i = 1; i < adapters.length; i++) {
    var currentJoltage = adapters[i];
    for (var diff in [1, 2, 3]) {
      ways[currentJoltage] = (ways[currentJoltage] ?? 0) + (ways[currentJoltage - diff] ?? 0);
    }
  }

  return ways[adapters.last] ?? 0;
}

File: swift/day10_part2_2020.swift

import Foundation

let fileURL = URL(fileURLWithPath: "input.txt")
let input = try String(contentsOf: fileURL)
let lines = input.components(separatedBy: .newlines)

var adapters = [0]
for line in lines {
    if let joltage = Int(line) {
        adapters.append(joltage)
    }
}

adapters.sort()
adapters.append(adapters.last! + 3)

func countArrangements(_ adapters: [Int]) -> Int64 {
    var ways: [Int: Int64] = [0: 1]

    for i in 1..<adapters.count {
        let currentJoltage = adapters[i]
        for diff in [1, 2, 3] {
            ways[currentJoltage, default: 0] += ways[currentJoltage - diff, default: 0]
        }
    }

    return ways[adapters.last!, default: 0]
}

print(countArrangements(adapters))
